### PR TITLE
fix quadratic behavior in natdynlink by using a STW section for frame-descriptor updates

### PR DESCRIPTION
**Pull request by gasche**
This PR proposes a fix for the quadratic behavior discussed in https://github.com/ocaml/ocaml/pull/11673#issuecomment-1297384255 (which causes performance regressions in both Frama-C and Coq, see also https://github.com/ocaml/ocaml/issues/11926).

We have discussed several implementation strategies to fix the quadratic behavior:

1. use the STW for synchronization: https://github.com/ocaml/ocaml/pull/11673#issuecomment-1399451455
2. use a hashtable with safe concurrent addition: https://github.com/ocaml/ocaml/issues/11926#issuecomment-1399512446
3. use a reader-writer lock: https://github.com/ocaml/ocaml/pull/11673#issuecomment-1399530867

This PR implements option (1): updates to the "current" caml_frame_descrs table are performed in a STW section. This allows unprotected access to caml_frame_descrs from mutators.

Note: I have not tested yet that the quadratic behavior is really gone from Frama-C or Coq. (Compiling them on the current trunk sounds a bit exhausting.) I am thinking of writing a synthetic test instead, but I am not sure how to proceed with that.

### CODE REVIEW

**Comment by kayceesrk** [comment]
> Note: I have not tested yet that the quadratic behavior is really gone from Frama-C or Coq. (Compiling them on the current trunk sounds a bit exhausting.) I am thinking of writing a synthetic test instead, but I am not sure how to proceed with that.

Would it be easier to backport this change to 5.0 and then get Coq and Frama-C to compile? I suspect that the changes between 5.0 and trunk in this part of the code are going to be minimal.

**Comment by gasche** [comment]
Some performance numbers on a synthetic benchmark, dynlinking 200 .cmxs each with 1000 small functions.

4.14: 1.0s
5.0: 1.7s
trunk: 1.4s
this PR: 1.0s

Good news: this PR does recover the 4.14 performance on this test. (In particular, running 200 stop-the-world has neglectible cost.)

Meh news: the cost of the test appears to be dominated by system-level stuff rather than building the hashtable, so I couldn't observe a nice quadratic plot turning into a linear plot.

**Comment by lthls** [comment]
I suspect that having 1000 small functions doesn't generate a lot of frametable entries (but I could be wrong).
A single function doing 1000 allocations or external calls might exercise it better (or not, I haven't actually tried).

**Comment by stedolan** [comment]
> I suspect that having 1000 small functions doesn't generate a lot of frametable entries (but I could be wrong).

This depends on what's in the small functions. Frametable entries are only generated at calls and allocations. In particular, `fun () -> 42` generates no frametable entries at all, but `fun () -> ref 42` generates one.

**Comment by gasche** [comment]
Each small function in my test contains a function call: the code looks like

```ocaml
let f0[@inline never] () = 0
let f1[@inline never] () = f1 () + 1
let f2[@inline never] () = f2 () + 2
...
```

I could of course tune my synthetic test, but I have the impression that what has been done already answers the question I was looking for: "does this restore 4.14 performance for repeated dynlinking?". The answer is: yes.

I'm thinking of porting a version of the patch (and the buildup PRs) on top of 5.0, so that Frama-C and Coq people can test it. But I think that interested people could already consider reviewing the PR now.

Edit: I added the `[@inline never]` in my example above that I forgot, answering @lthls' question below. If people want to run their own synthetic benchmarks, here is my program-generation script:

```ocaml
let n = try int_of_string Sys.argv.(1) with _ ->
  prerr_endline "Expected argument: number of files to generate";
  exit 1

let () =
  for i = 1 to n do
    let fname = Printf.sprintf "test%04d.ml" i in
    Out_channel.with_open_text fname begin fun oc ->
      Printf.fprintf oc
        "let[@inline never] f%d () = 0\n%!"
        0;
      for j = 1 to 1000 do
        Printf.fprintf oc
          "let[@inline never] f%d () = f%d () + 1\n%!"
          j (j - 1)
      done;
      Printf.fprintf oc
        "\nlet () = print_int (f%d ()); print_newline ()\n%!"
        (i mod 1000)
    end
  done
```

and my benchmark:

```ocaml
let n = try int_of_string Sys.argv.(1) with _ ->
  prerr_endline "Expected argument: number of files to dynlink";
  exit 1

let () =
  for i = 1 to n do
    let cmxs = Printf.sprintf "test%04d.cmxs" i in
    Dynlink.loadfile cmxs
  done
```

**Comment by lthls** [comment]
If your code is exactly like that, then inlining will transform all your functions into constants and you will have more or less zero frametable entries. You might want to check the generated Cmm of your code to make sure the calls are still there.

**Comment by gadmm** [comment]
A benchmark is not necessary to know that the strategy in this PR will result in comparable performance as 4.14 when a single domain is involved. The big unknown is the behaviour in real workloads with multiple domains.

This is the second PR (with Stephen's) that tries to provide a partial fix to the issue, while I have received no indication that the total fix I proposed is something that interests you (and by seeing several partial fixes being worked on, I am led to believe that it actually does not interest you). It might be useful to discuss what is your desired end result before working on specific solutions.

**Comment by gasche** [comment]
I am not aware of existing programs that load plugins *in parallel* with heavy computations in other domains.

> This is the second PR (with Stephen's) that tries to provide a partial fix to the issue, while I have received no indication that the total fix I proposed is something that interests you (and by seeing several partial fixes being worked on, I am led to believe that it actually does not interest you).

Is this a passive-agressive way to say that I should be working on your proposed solution to the performance regression, rather than my proposed solution to the performance regression? Feel free to send a PR.

Note that an approach that improves scalability could always be proposed later -- it is much easier to do on top of this PR than on top of the trunk as it was before Stephen and myself started working on this.

One advantage of my approach is that it is obvious that there is no cost for programs that only access the frametables, without using Dynlink. With either your approach of the reader-writer mutex proposed by Xavier, we would want to benchmark this to ensure that the synchronization costs for reader do not add overhead. This sounds like work, and I'm too lazy to do it, so I prefer working on the STW approach myself. This is not at all a criticism of both your suggestions, which I think are good ideas as well.

> It might be useful to discuss what is your desired end result before working on specific solutions.

My desired end-result is to solve the performance regression that affects Frama-C and Coq.

**Comment by gadmm** [comment]
> Is this a passive-agressive way to say that I should be working on your proposed solution to the performance regression, rather than my proposed solution to the performance regression? Feel free to send a PR.

No, it says what it says. You can understand that I am not going to work on it if I see you working on the same problem at the same time; in particular if you think you are happy with the end result of this PR then you are unlikely to see a PR for my approach. It is not a race, there can be time for discussion of what is the best approach in a way that minimizes collective work.

I do not think that benchmarks are needed in my approach.

**Comment by gasche** [comment]
I have rebased this PR on top of 5.0 in my remote branch
  https://github.com/gasche/ocaml/commits/frametables-in-stw-5.0

(There are minor conflicts with other changes in trunk, so it is not fully automatic.)

If people interested in Frama-C or Coq would like to try it, they could see whether the dynlink-related slowdown are fixed for their use-case.

cc @maroneze @silene

**Comment by gasche** [comment]
Note: to test a branch, you probably want to clone the repository+branch and create an opam switch for it, which is described in [HACKING.adoc#opam-switch](https://github.com/ocaml/ocaml/blob/5.0.0/HACKING.adoc#opam-switch).

**Comment by gasche** [comment]
@lpw25 points out that the need to fix the quadratic behavior may have vanished in practice after Stephen's batching change in #11935. We should consider sending only that change to Frama-C/Coq people to see.

**Comment by gasche** [comment]
I pushed a 5.0 branch with only the "batch natdynlink calls" change from Stephen:

  https://github.com/gasche/ocaml/tree/natdynlink-batching-5.0

This would also be a candidate to see whether it solves the performance regression in practice. (It is a prefix of the other branch. It does not make the quadratic slowdown go away, but it reduces the number of iterations that give this quadratic behavior by a large constant factor, so it may be enough in practice.)

**Comment by maroneze** [comment]
Thanks for all the effort and the documentation about compiling a git branch via `opam switch install .`, it's very useful.

I tried cloning your repository and compiling your branch (32e41b9a877d97d3e856cca6d4c5afb62098090e), and ran the OSCS benchmarks, and they seem *extremely* close to those of 4.14.1 (no discernible difference, at least for a first run). Which makes me wonder if I did the right tests.

For instance, the `VERSION` file does contain `5.0.1+dev0-2022-12-15`, but `ocaml --version` in the switch created from `opam switch create . --empty` still outputs `4.14.1`.

Is there something else I could try to make sure I did the right tests?

I'm going to benchmark this branch ([a2c5522](https://github.com/gasche/ocaml/commit/a2c55223899a82179f3bf8a2c265d09d8b55f70d)) and see if I get something different.

**Comment by maroneze** [comment]
Well, after trying with the other branch, I also got the same times, so I'm definitely running the 4.14 compiler, sorry about the noise.

I just tried erasing some of my switches that were related to `ocaml-variants`, and using the other commands you indicate, and it seems to be actually using the right version this time.

I'll try running the `natdynlink-batching-5.0` branch.

However, for this branch, `frametables-in-stw`, it seems it needs 5.1.0 (instead of 5.0.1), and due to the lack of packages such as `ppx_import`, which are required by Frama-C, I am unable to compile it. Is there a way to compile this branch with a 5.0.1 variant?

**Edit**: nevermind, I just saw the other message with the 5.0 branch, I'll benchmark it and report later.

**Comment by gasche** [comment]
Thanks @maroneze for the tests! `frametables-in-stw-5.0` is the version that is backported on top of 5.0.

**Comment by silene** [comment]
See #11926 for the methodology and the original results.

TLDR, this pull request brings a 4% speedup (on top of the 17% of the other pull request).

With  https://github.com/gasche/ocaml/tree/frametables-in-stw-5.0
```
        282 319,05 msec task-clock:u              #    1,015 CPUs utilized          
                 0      context-switches:u        #    0,000 /sec                   
                 0      cpu-migrations:u          #    0,000 /sec                   
        40 002 642      page-faults:u             #  141,693 K/sec                  
 1 127 417 356 991      cycles:u                  #    3,993 GHz                    
 1 675 032 758 744      instructions:u            #    1,49  insn per cycle         
   354 410 147 630      branches:u                #    1,255 G/sec                  
     8 488 399 007      branch-misses:u           #    2,40% of all branches        

     278,090636551 seconds time elapsed

     245,817631000 seconds user
      36,605027000 seconds sys
```

With https://github.com/gasche/ocaml/tree/natdynlink-batching-5.0

```
        293 496,99 msec task-clock:u              #    1,014 CPUs utilized          
                 0      context-switches:u        #    0,000 /sec                   
                 0      cpu-migrations:u          #    0,000 /sec                   
        40 763 930      page-faults:u             #  138,890 K/sec                  
 1 163 996 708 051      cycles:u                  #    3,966 GHz                    
 1 724 278 950 322      instructions:u            #    1,48  insn per cycle         
   364 935 195 285      branches:u                #    1,243 G/sec                  
     8 837 654 424      branch-misses:u           #    2,42% of all branches        

     289,330869022 seconds time elapsed

     255,717521000 seconds user
      37,884676000 seconds sys
```

Since the frequency of the processor does not correspond to the one from the previous benchmark for some reason (this is not a laptop), here are new runs for the baseline, to be on the safe side.

With 4.14.1:

```
        274 895,31 msec task-clock:u              #    1,014 CPUs utilized          
                 0      context-switches:u        #    0,000 /sec                   
                 0      cpu-migrations:u          #    0,000 /sec                   
        38 798 237      page-faults:u             #  141,138 K/sec                  
 1 075 218 839 586      cycles:u                  #    3,911 GHz                    
 1 508 380 673 624      instructions:u            #    1,40  insn per cycle         
   317 312 647 833      branches:u                #    1,154 G/sec                  
     8 938 167 348      branch-misses:u           #    2,82% of all branches        

     271,072333982 seconds time elapsed

     239,020692000 seconds user
      35,974261000 seconds sys
```

With 5.0.0:

```
        347 458,80 msec task-clock:u              #    1,012 CPUs utilized          
                 0      context-switches:u        #    0,000 /sec                   
                 0      cpu-migrations:u          #    0,000 /sec                   
        44 326 327      page-faults:u             #  127,573 K/sec                  
 1 385 205 826 883      cycles:u                  #    3,987 GHz                    
 1 988 689 256 937      instructions:u            #    1,44  insn per cycle         
   422 048 922 722      branches:u                #    1,215 G/sec                  
    11 879 729 454      branch-misses:u           #    2,81% of all branches        

     343,295835599 seconds time elapsed

     306,331335000 seconds user
      41,240759000 seconds sys
```

**Comment by maroneze** [comment]
Please take this with a large grain of salt, but I think this time I did bench the correct branches.

The screenshot below summarizes the differences (a baseline run on 4.14, another with 5.0.1 from Feb 1st, then the two branches); the numbers in the last lines correspond to the *average* and *median* values of the rows above them:

Screenshot_20230208_125958

So, to summarize: from 4.14 to 5.0.1, there was an increase of 31% in execution time, lowered to 13% with natdylink-batching, and 9% with frametables-in-stw.

The raw data is in the file below (note: it also contains other runs which are not relevant, such as several runs for 4.14), with a pivot table.

[oscs-comparison.xlsx](https://github.com/ocaml/ocaml/files/10685202/oscs-comparison.xlsx)

The numbers seem not too far from those of @silene, which makes me believe I didn't mess them up.

**Edit**: after reading a bit more carefully, it seems that the overhead for the long running benchmarks has changed llittle between them; at least for one of Frama-C's plugins (Eva), this is the metric that impacts industrial users the most, so the overhead is still significant.

Screenshot_20230208_134856

**Comment by gasche** [comment]
Thanks both! My summary of your results:

- as @lpw25 suspected, the "batching" change (moving from a N^2 algorithm to a (N/10)^2) removes a large part of the slowdown: on the Coq measurements, this brings a 17% speedup against the current trunk
- but there remains a measurable benefit to turning the (N/10)^2 algorithm into an N/10 algorithm: on the Coq measurements, this brings a 4% speedup on top of the previous speedup
- in other words both changes are necessary to recover 4.14-like performance for plugins; for example the Frama-C 2048 examples is 10% slower than 4.14 with just batching, and 1% slower with both changes (that is, same as 4.14 within noie)
- of course there may remain other slowdown factors that affect Coq and Frama-C that are not related to dynlink

> it seems that the overhead for the long running benchmarks has changed llittle between them

Of course, the changes we are discussing only affect the performance of Dynlink, that is, loading plugins. Assuming most programs only load plugins at the beginning, this only affects the startup time -- which is significant for short-running programs.

**Comment by gasche** [comment]
Given than the fix proposed in the present PR is fairly simple, and it does make a measurable difference compared to the current trunk (which already has Stephen's batching change), I would encourage people to review it with the objective of merging it. (Again: alternative approaches have been proposed, but it's not really a one-or-another choice, the PR proposed here also makes these alternative approaches easier to implement.)

**Comment by gasche** [comment]
Still interested in finding reviewers for this PR. Would one of @sadiqj, @Engil, @lthls or @fabbing be interested? This is a relatively small patch (40 added lines, 97 removed lines), and it can be reviewed commit-by-commit.

**Comment by sadiqj** [comment]
> Still interested in finding reviewers for this PR. Would one of @sadiqj, @Engil, @lthls or @fabbing be interested? This is a relatively small patch (40 added lines, 97 removed lines), and it can be reviewed commit-by-commit.

Thanks for the ping @gasche . I can take a look at this tomorrow.

**Review by sadiqj — APPROVED** [review]
This looks good to me

One thing to note is that there's now a poll point in `caml_register_frametables` where there wasn't before. I had a look at `caml_natdynlink_register` and I think this is safe.

**Comment by gasche** [comment]
Thanks @sadiqj for your review! Good point about the poll point; I looked at caml_natdynlink_register again and I agree that it seems to be robust against having a GC run the middle.

I added a Changes entry and I think this is good to merge once the CI agrees.
